### PR TITLE
fix: timestamp formatting in defaultValue

### PR DESCRIPTION
**Pull request by SwaySway**
*Issue #, if available:*
*Description of changes:*
- handle defaultValues date associated fields
  - for `AWSTimestamp` this will convert the epoch seconds/milliseconds to datetime-local input
  - for `AWSDateTime` will remove the 'Z' character so the date is laid out correctly in the input field

note: seconds/milliseconds are not editable
### format for epoch in seconds
Screen Shot 2022-09-26 at 1 48 03 PM

### format for epoch in milliseconds
Screen Shot 2022-09-26 at 1 49 05 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/codegen-ui-react/lib/utils/forms/value-mappers.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+/*
+given a timestamp number it will convert it to datetime local for forms
+first does a check to see if the timestamp is in seconds or milliseconds
+*/
+export const getDateTimeLocal = (ts: number) => {
+  if (Math.abs(Date.now() - ts) < Math.abs(Date.now() - ts * 1000)) {
+    return new Date(ts).toISOString().replace('Z', '');
+  }
+  return new Date(ts * 1000).toISOString().replace('Z', '');
```
Comment (awinberg-aws) [inline packages/codegen-ui-react/lib/utils/forms/value-mappers.ts:27]: this isn't a "local" conversion so much as it's just stripping off the UTC marker. Why was this necessary, and are there any data integrity concerns here?

Comment (SwaySway) [reply packages/codegen-ui-react/lib/utils/forms/value-mappers.ts:27]: This change is added so the date timestamp values being passed in are viewable in an update form.
This shouldn't change the date value itself as its only being used in defaultValue.
Not opposed to not supporting the defaultValue is there a concern where we don't want this value mapper?

**Review by hein-j — DISMISSED** [review]

File: packages/codegen-ui-react/lib/forms/component-helper.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+import { FieldConfigMetadata } from '@aws-amplify/codegen-ui';
+import { factory, SyntaxKind } from 'typescript';
+
+export const ControlledComponents = ['StepperField', 'SliderField', 'SelectField'];
```
Comment (letsbelopez) [inline packages/codegen-ui-react/lib/forms/component-helper.ts:20]: SwitchField and ToggleButton are most likely controlled as well. What's the impact of not including all controlled components?

File: packages/codegen-ui-react/lib/forms/component-helper.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+          factory.createIdentifier(stateName),
+          factory.createIdentifier('replace'),
+        ),
+        undefined,
+        [factory.createStringLiteral('Z'), factory.createStringLiteral('')],
+      ),
+    );
+  }
+
+  return factory.createJsxAttribute(factory.createIdentifier('defaultValue'), expression);
```
Comment (letsbelopez) [inline packages/codegen-ui-react/lib/forms/component-helper.ts:66]: I think the defaultValue attribute name is going to change per component type. I think we need a map of componenttype to default value attribute.

For example,

File: packages/codegen-ui-react/lib/utils/forms/value-mappers.ts (hunk 1 above)
Comment (SwaySway) [reply packages/codegen-ui-react/lib/utils/forms/value-mappers.ts]: synced offline - reformatting the date so the local time is preserved only when presenting the value

File: packages/codegen-ui-react/lib/forms/component-helper.ts (hunk 3 above)
Comment (SwaySway) [reply packages/codegen-ui-react/lib/forms/component-helper.ts:64]: adding a map in the next revision

**Review by awinberg-aws — APPROVED** [review]

File: packages/codegen-ui-react/lib/forms/component-helper.ts (hunk 2 above)
Comment (SwaySway) [reply packages/codegen-ui-react/lib/forms/component-helper.ts:20]: the impact is minimal though the idea is to not render default values for these controlled components as the value is coming from the state - will add the other two

File: packages/codegen-ui-react/lib/utils/forms/value-mappers.ts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    year: 'numeric',
+    month: '2-digit',
+    day: '2-digit',
+    hour: '2-digit',
+    minute: '2-digit',
+    second: '2-digit',
+    fractionalSecondDigits: 3,
+  });
+  const parts = df.formatToParts(date).map((v) => v.value);
+  return `${parts[4]}-${parts[0]}-${parts[2]}T${parts[6]}:${parts[8]}:${parts[10]}.${parts[12]}`;
```
Comment (awinberg-aws) [inline packages/codegen-ui-react/lib/utils/forms/value-mappers.ts:50]: This will not work for PM, `df.formatToParts(new Date('Mon Sep 26 2022 20:55:58 GMT-0700 (Pacific Daylight Time)'))` gives `parts[8] === "08"` instead of `"20"`

**Review by awinberg-aws — CHANGES_REQUESTED** [review]

File: packages/codegen-ui-react/lib/utils/forms/value-mappers.ts (hunk 4 above)
Comment (awinberg-aws) [reply packages/codegen-ui-react/lib/utils/forms/value-mappers.ts:50]: Adding `hour12: false` to the config object above will address the AM/PM issue but another thing to be aware of is that the order of parts also changes depending on the locale used by DateTimeFormat. Not sure if `'default'` is meant to vary by the user's actual locale or not but that could work against us for this implementation.

Comment (SwaySway) [reply packages/codegen-ui-react/lib/utils/forms/value-mappers.ts:50]: new rev
- changing to set hour12 to false
- for the times we can just index by the key as we know thats a constant for the response from date time format

**Review by cwoolum — APPROVED** [review]
